File: Demos/OC/UFileSDKDemo/UFileSDKDemo/setting/QRCodeVC.m
//
//  QRCodeVC.m
//  UFileSDKDemo
//
//  Created by ethan on 2018/12/5.
//  Copyright © 2018 ucloud. All rights reserved.
//

#import "QRCodeVC.h"
#import <AVFoundation/AVFoundation.h>
#import "UFileSDKDemoConst.h"
#import "DataTools.h"

@interface QRCodeVC ()<AVCaptureMetadataOutputObjectsDelegate>
@property (strong, nonatomic) IBOutlet UIView *scanView;

@property (nonatomic,strong) AVCaptureDevice *m_device;
@property (nonatomic,strong) AVCaptureSession *m_session;
@property (nonatomic,strong) AVCaptureMetadataOutput *metadata_output;
@property (nonatomic,strong) AVCaptureVideoPreviewLayer *m_previewLayer;

@end

@implementation QRCodeVC

- (AVCaptureSession *)m_session
{
    if (!_m_session) {
        NSError *error;
        _m_device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
        AVCaptureDeviceInput *input =[AVCaptureDeviceInput deviceInputWithDevice:_m_device error:&error];
        if (!input) {
            NSLog(@"%@",[error localizedDescription]);
        }
        _m_session = [[AVCaptureSession alloc] init];
        if ([_m_session canAddInput:input]) {
            [_m_session addInput:input];
        }
        _metadata_output = [[AVCaptureMetadataOutput alloc] init];
        if ([_m_session canAddOutput:_metadata_output]) {
            [_m_session addOutput:_metadata_output];
        }
        
        dispatch_queue_t videoQueue = dispatch_queue_create("videoQueue", NULL);
        [_metadata_output setMetadataObjectsDelegate:self queue:videoQueue];
        
        // 设置元数据类型
        [_metadata_output setMetadataObjectTypes:@[AVMetadataObjectTypeQRCode]];
        
        _m_previewLayer = [[AVCaptureVideoPreviewLayer alloc] initWithSession:_m_session];
        [_m_previewLayer setVideoGravity:AVLayerVideoGravityResizeAspectFill];
        [_m_previewLayer setFrame:self.scanView.layer.bounds];
        [self.scanView.layer insertSublayer:_m_previewLayer atIndex:0];
    }
    return _m_session;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self startPreview];
}


- (void)startPreview
{
    if (![self.m_session isRunning]) {
        [self.m_session startRunning];
    }
}

- (void)stopPreview
{
    if ([self.m_session isRunning]) {
        [self.m_session stopRunning];
    }
}

#pragma mark -AVCaptureMetadataOutputObjectsDelegate
- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection
{
    if (metadataObjects == nil || [metadataObjects count] <= 0) {
        return;
    }
    
    AVMetadataMachineReadableCodeObject *metadataObj = [metadataObjects objectAtIndex:0];
    NSString *result;
    if ([[metadataObj type] isEqualToString:AVMetadataObjectTypeQRCode]) {
        result = metadataObj.stringValue;
    }else{
        NSLog(@"Your scanning is not QR code..");
        return;
    }
    [self stopPreview];
    result = metadataObj.stringValue;
    [self performSelectorOnMainThread:@selector(dealwithScanResult:) withObject:result waitUntilDone:NO];
   
}

- (void)dealwithScanResult:(NSString *)result
{
    __weak typeof(self) weakself = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakself stopPreview];
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"扫描结果" message:result preferredStyle:(UIAlertControllerStyleAlert)];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"填充" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [weakself parseJsonStrAndStore:result];
            [weakself.navigationController popViewControllerAnimated:YES];
        }];
        [alert addAction:okAction];
        
        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"重扫" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            [weakself startPreview];
        }];
        [alert addAction:cancelAction];
        
        [weakself presentViewController:alert animated:YES completion:nil];
    });
    NSLog(@"%@",result);
    
}

- (void)parseJsonStrAndStore:(NSString *)jsonStr
{
    if (!jsonStr) {
        NSLog(@"scan QR Code, jsonStr is nil..");
        return;
    }

    NSData *jsonData = [jsonStr dataUsingEncoding:NSUTF8StringEncoding];
    NSError *jsonError;
    NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingMutableContainers error:&jsonError];
    if (jsonError) {
        NSLog(@"scan QR Code, parse jsonStr error ,error info-->%@",jsonError.description);
        return;
    }
    
    if (dict) {
        NSArray *keys = @[KBucketPublicKey,KBucketPrivateKey,KProfixSuffix,KBucketName,KFileOperateEncryptServer,KFileAddressEncryptServer];
        for (NSString *key in keys) {
            [DataTools storeStrData:[dict objectForKey:key] keyName:key];
        }
    }
        

}

@end

File: Demos/Swift/UFileSDKDemo-swift/UFileSDKDemo-swift/setting/QRCodeVC.swift
//
//  QRCodeVC.swift
//  UFileSDKDemo-swift
//
//  Created by ethan on 2018/12/13.
//  Copyright © 2018 ucloud. All rights reserved.
//

import UIKit
import AVFoundation

class QRCodeVC: UIViewController,AVCaptureMetadataOutputObjectsDelegate{

    @IBOutlet weak var scanView:UIView!
    
    var m_device:AVCaptureDevice!
    var m_session:AVCaptureSession!
    var metadata_output:AVCaptureMetadataOutput!
    var m_previewLayer:AVCaptureVideoPreviewLayer!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        m_device = AVCaptureDevice.default(for: .video)
        do {
           let input:AVCaptureDeviceInput =  try AVCaptureDeviceInput.init(device: m_device)
           m_session = AVCaptureSession.init()
            if m_session.canAddInput(input){
                m_session.addInput(input)
            }
            
            metadata_output = AVCaptureMetadataOutput.init()
            if m_session.canAddOutput(metadata_output) {
                m_session .addOutput(metadata_output)
            }
            
            let serialQueue  = DispatchQueue(label: "videoQueue")
            metadata_output.setMetadataObjectsDelegate(self, queue: serialQueue)
            
            metadata_output.metadataObjectTypes = [AVMetadataObject.ObjectType.qr]
            
            m_previewLayer = AVCaptureVideoPreviewLayer.init(session: m_session)
            m_previewLayer.videoGravity = AVLayerVideoGravity.resizeAspectFill
            m_previewLayer.frame = self.scanView.layer.bounds
            self.scanView.layer.insertSublayer(m_previewLayer, at: 0)
            
        } catch let error as NSError  {
            print(error.description)
        }
        
        self.startPreview()
        
    }
    
    func startPreview() -> Void {
        if self.m_session.isRunning == false {
            self.m_session.startRunning()
        }
    }
    
    func stopPreview() -> Void {
        if self.m_session.isRunning {
            self.m_session.stopRunning()
        }
    }
    
    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        if metadataObjects == nil || metadataObjects.count <= 0 {
            return
        }
        
        let metadataObj:AVMetadataMachineReadableCodeObject = metadataObjects[0] as! AVMetadataMachineReadableCodeObject
        var result:String
        if metadataObj.type == AVMetadataObject.ObjectType.qr {
            result = metadataObj.stringValue!
        }else {
            print("Your scaning is not QR code..")
            return
        }
        self.stopPreview()
        result = metadataObj.stringValue!
        
        DispatchQueue.main.async {
            self.dealwithScanResult(result: result)
        }
        
//        self.performSelector(onMainThread:Selector(("dealwithScanResult")), with: result, waitUntilDone: false)
    }
    
    func dealwithScanResult(result:String) -> Void {
        weak var weakself:QRCodeVC! = self
        DispatchQueue.main.async {
            weakself.stopPreview()
            let alert:UIAlertController = UIAlertController.init(title: "扫描结果", message: result, preferredStyle: .alert)
            let okAction = UIAlertAction.init(title: "填充", style: .default, handler: { (UIAlertAction) in
                weakself.parseJsonStrAndStore(jsonStr: result)
                weakself.navigationController?.popViewController(animated: true)
                
            })
            alert.addAction(okAction)
            
            let cancelAction = UIAlertAction.init(title: "重扫", style: .cancel, handler: { (UIAlertActio) in
                weakself.startPreview()
            })
            
            alert.addAction(cancelAction)
            weakself.present(alert, animated: true, completion: nil)
        }
        print(result)
    }
    
    func parseJsonStrAndStore(jsonStr:String) -> Void {
        if jsonStr == nil {
           print("scan QR Code , jsonStr is nil..")
            return
        }
        
        let jsonData:Data = jsonStr.data(using: .utf8)!
        
        do {
            let dict:NSDictionary = try JSONSerialization.jsonObject(with:jsonData as Data, options: .mutableContainers) as! NSDictionary
            let keys = [KBucketPublicKey,KBucketPrivateKey,KProfixSuffix,KBucketName,KFileOperateEncryptServer,KFileAddressEncryptServer]
            for key in keys {
                DataTools.storeStrData(strData: dict.object(forKey: key) as! String, and: key)
            }
            
        } catch let error as NSError {
            print(error.description)
        }
        
    }



}
